File: src/medium/t4/solve.go
package main

import "fmt"

func maxArea(height []int) (max int) {
	var minH = 0

	for i, j := 0, len(height)-1; i < j; {
		weight := j - i
		if height[i] < height[j] {
			minH = height[i]
			i++
		} else {
			minH = height[j]
			j--
		}
		if weight*minH > max {
			max = weight * minH
		}
	}
	return
}

func main() {
	var height = []int{1, 8, 6, 2, 5, 4, 8, 3, 7}
	fmt.Print(maxArea(height))
}

File: src/medium/t4/solve.rs
pub fn max_area(height: Vec<i32>) -> i32 {
    let mut max = 0;
    let mut i: usize = 0;
    let mut j = height.len() - 1;
    let mut weight = 0;
    while i < j {
        weight = j - i;
        let min_h = {
            if height[i] < height[j] {
                i += 1;
                height[i - 1]
            } else {
                j-=1;
                height[j+1]
            }
        };
        max = match weight as i32 * min_h {
            tmp if tmp > max => tmp,
            _ => max,
        };
    }
    max
}
